File: plugin/src/main/com/suhininalex/clones/Token.java
package com.suhininalex.clones;

import com.intellij.psi.PsiElement;
import org.jetbrains.annotations.NotNull;

public class Token implements Comparable<Token> {

    /**
     * Position in code
     */
    public PsiElement source;

    /**
     * Token id
     */
    //TODO finals
    public  Short value;
    private  int position;

    public int getPosition() {
        return position;
    }

    public Token(PsiElement source, int position) {
        this.source = source;
        this.position = position;
        value = source.getNode().getElementType().getIndex();
    }

    //TODO delete this
    public Token(int position, short value){
        this.position = position;
        this.value = value;
    }

    @NotNull
    Short getValue(){
        return value;
    }

    @Override
    public int hashCode() {
        return getValue().hashCode();
    }

    @Override
    public  int compareTo(Token o){
        return getValue().compareTo(o.getValue());
    }

    @Override
    public boolean equals(Object obj) {
        if (obj instanceof Token)
            return getValue().equals(((Token) obj).getValue());
        else return false;
    }

    @Override
    public String toString() {
        return getValue().toString();//source.getNode().getElementType().toString();
//        return "("+value + ":"+source.getText()+")";
    }
}

File: plugin/src/main/com/suhininalex/clones/Token.kt
package com.suhininalex.clones

import com.intellij.psi.PsiElement
import com.intellij.psi.PsiMethod
import com.intellij.psi.impl.source.tree.ElementType
import com.intellij.psi.tree.TokenSet
import net.suhininalex.kotlin.utils.abbrevate

class Token(val source: PsiElement, val method: PsiMethod) : Comparable<Token> {

    val nonAnonimized = source.notNeedAnonimization()

    val value: Int =
        if (nonAnonimized)
            source.text.hashCode()
        else source.node.elementType.index.toInt()

    override fun hashCode() = value.hashCode()

    override fun compareTo(other: Token) = value.compareTo(other.value)

    override fun equals(other: Any?) =
        if (other is Token) value == other.value
        else false

    override fun toString() =
        if (nonAnonimized) source.text.abbrevate(5)
        else source.node.elementType.toString()
}

fun PsiElement.notNeedAnonimization() =
    parent.parent in TokenSet.create(ElementType.METHOD_CALL_EXPRESSION) &&
    parent in TokenSet.create(ElementType.REFERENCE_EXPRESSION) &&
    this in TokenSet.create(ElementType.IDENTIFIER)